lib/Transforms/PrintOpCount.cpp

```diff
@@ -0,0 +1,111 @@
+//===- PrintOpCount.cpp - Operation Count Emission Pass ---------*- C++ -*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// Contains a pass to emit operation count results
+//
+//
+//===----------------------------------------------------------------------===//
+
+#include "circt/Analysis/OpCountAnalysis.h"
+#include "circt/Transforms/Passes.h"
+#include "mlir/Pass/Pass.h"
+
+namespace circt {
+#define GEN_PASS_DEF_PRINTOPCOUNT
+#include "circt/Transforms/Passes.h.inc"
+} // namespace circt
+
+using namespace mlir;
+using namespace circt;
+
+namespace {
+
+void printOpAndOperandCounts(analysis::OpCountAnalysis &opCount,
+                             raw_ostream &os, bool sorted = false) {
+  auto opNames = opCount.getFoundOpNames();
+  // Sort to account for non-deterministic DenseMap ordering
+  if (sorted)
+    llvm::sort(opNames, [](OperationName name1, OperationName name2) {
+      return name1.getStringRef() < name2.getStringRef();
+    });
+  for (auto opName : opNames) {
+    os << "- name: " << opName << "\n";
+    os << "  count: " << opCount.getOpCount(opName) << "\n";
+    auto operandMap = opCount.getOperandCountMap(opName);
+    if (operandMap.size() <= 1)
+      continue;
+    llvm::SmallVector<size_t> keys;
+    for (auto pair : operandMap)
+      keys.push_back(pair.first);
+    // Sort for determinism if required again
+    if (sorted)
+      llvm::sort(keys);
+    for (auto num : keys) {
+      os << "    - operands: " << num << "\n";
+      os << "      count: " << operandMap[num] << "\n";
+    }
+  }
+}
+
+void printOpAndOperandJSON(analysis::OpCountAnalysis &opCount,
+                           raw_ostream &os) {
+  auto opNames = opCount.getFoundOpNames();
+  os << "{\n";
+  for (auto [i, opName] : llvm::enumerate(opNames)) {
+    os << " \"" << opName << "\": {\n";
+    auto operandMap = opCount.getOperandCountMap(opName);
+    for (auto [j, pair] : llvm::enumerate(operandMap)) {
+      os << "  " << pair.first << ": " << pair.second;
+      if (j != operandMap.size() - 1)
+        os << ",";
+      os << "\n";
+    }
+    os << " }";
+    if (i != opNames.size() - 1)
+      os << ",";
+    os << "\n";
+  }
+  os << "}\n";
+}
```

Comment:
This would benefit from use of `llvm::json` to avoid having to hand-roll JSON emission.

An example of a very compact usage of that API: https://github.com/llvm/llvm-project/blob/09ba83be0ac178851e3c9c9c8fefddbdd4d8353f/llvm/lib/Analysis/TensorSpec.cpp#L50

Reply:
Good point, this was a pretty easy change so I've updated this PR to use it